File: golang/Python3/GoParserBase.py
import sys
from antlr4 import *
if sys.version_info[1] > 5:
    from typing import TextIO
else:
    from typing.io import TextIO

class GoParserBase(Parser):

    debug = False

    @staticmethod
    def _has_arg(args, arg):
        arg_lower = arg.lower()
        for a in args:
            if arg_lower in a.lower():
                return True
        return False

    def __init__(self, input:TokenStream, output:TextIO = sys.stdout):
        super().__init__(input, output)
        self.table = set()
        GoParserBase.debug = GoParserBase._has_arg(sys.argv, '--debug')
        if GoParserBase.debug:
            print("debug =", GoParserBase.debug)

    def myreset(self):
        self.table = set()
    
    def closingBracket(self) -> bool:
        la = self._input.LT(1)
        return la.type == self.R_PAREN or la.type == self.R_CURLY or la.type == Token.EOF

    def isNotReceive(self) -> bool:
        la = self._input.LT(2)
        return la.type != self.RECEIVE

    def addImportSpec(self):
        ctx = self._ctx
        if not isinstance(ctx, self.ImportSpecContext):
            return
        importSpec = ctx
        if importSpec.packageName() is not None:
            name = importSpec.packageName().getText()
            if self.debug:
                print("Entering", name)
            self.table.add(name)
            return
        import_path = importSpec.importPath()
        if import_path is None:
            return
        name = import_path.getText()
        if self.debug:
            print("import path", name)
        name = name.replace("\"", "")
        if len(name) == 0:
            return
        name = name.replace("\\", "/")
        path_arr = name.split('/')
        if len(path_arr) == 0:
            return
        last_component = path_arr[-1]
        if len(last_component) == 0:
            return
        # Handle special cases like "." and ".."
        if last_component == "." or last_component == "..":
            return
        file_arr = last_component.split('.')
        # Guard against empty array (can happen if last_component is all dots)
        if len(file_arr) == 0:
            self.table.add(last_component)
            if self.debug:
                print("Entering", last_component)
            return
        file_name = file_arr[-1]
        if len(file_name) == 0:
            # Fall back to last_component if split resulted in empty string
            file_name = last_component
        if self.debug:
            print("Entering", file_name)
        self.table.add(file_name)

    def isOperand(self) -> bool:
        la = self._input.LT(1)
        if la.text == "err":
            return True
        result = True
        if la.type != self.IDENTIFIER:
            if self.debug:
                print(f"isOperand Returning {result} for {la.text}")
            return result
        result = (la.text in self.table)
        la2 = self._input.LT(2)
        if la2.type != self.DOT:
            result = True
            if self.debug:
                print(f"isOperand Returning {result} for {la.text}")
            return result
        la3 = self._input.LT(3)
        if la3.type == self.L_PAREN:
            result = True
            if self.debug:
                print(f"isOperand Returning {result} for {la.text}")
            return result
        if self.debug:
            print(f"isOperand Returning {result} for {la.text}")
        return result

    def isConversion(self) -> bool:
        la = self._input.LT(1)
        result = (la.type != self.IDENTIFIER)
        if self.debug:
            print(f"isConversion Returning {result} for {la.text}")
        return result

    def isMethodExpr(self) -> bool:
        la = self._input.LT(1)
        result = True
        if la.type == self.STAR:
            if self.debug:
                print(f"isMethodExpr Returning {result} for {la.text}")
            return result
        if la.type != self.IDENTIFIER:
            result = False
            if self.debug:
                print(f"isMethodExpr Returning {result} for {la.text}")
            return result
        result = (la.text not in self.table)
        if self.debug:
            print(f"isMethodExpr Returning {result} for {la.text}")
        return result

    # Built-in functions that take a type as first argument
    BUILTIN_TYPE_FUNCTIONS = {"make", "new"}

    # Check if we're in a call to a built-in function that takes a type as first argument.
    # Called after L_PAREN has been matched in the arguments rule.
    def isTypeArgument(self) -> bool:
        # After matching L_PAREN, LT(-1) is '(' and LT(-2) is the token before it
        func_token = self._input.LT(-2)
        if func_token is None or func_token.type != self.IDENTIFIER:
            if self.debug:
                print("isTypeArgument Returning False - no identifier before (")
            return False
        result = func_token.text in self.BUILTIN_TYPE_FUNCTIONS
        if self.debug:
            print(f"isTypeArgument Returning {result} for {func_token.text}")
        return result

    # Check if we're NOT in a call to a built-in function that takes a type.
    # This is the inverse of isTypeArgument for the expressionList alternative.
    def isExpressionArgument(self) -> bool:
        result = not self.isTypeArgument()
        if self.debug:
            print(f"isExpressionArgument Returning {result}")
        return result

File: golang/Cpp/GoParserBase.cpp
#include "GoParserBase.h"
#include "GoParser.h"
#include <algorithm>
#include <cctype>

bool GoParserBase::hasArg(int argc, char* argv[], const std::string& arg)
{
    std::string argLower = arg;
    std::transform(argLower.begin(), argLower.end(), argLower.begin(),
                   [](unsigned char c){ return std::tolower(c); });
    for (int i = 0; i < argc; i++) {
        std::string a = argv[i];
        std::transform(a.begin(), a.end(), a.begin(),
                       [](unsigned char c){ return std::tolower(c); });
        if (a.find(argLower) != std::string::npos) {
            return true;
        }
    }
    return false;
}

GoParserBase::GoParserBase(antlr4::TokenStream* input, int argc, char* argv[])
    : Parser(input)
{
    debug = hasArg(argc, argv, "--debug");
    if (debug) {
        std::cout << "debug = " << debug << std::endl;
    }
}

void GoParserBase::myreset()
{
    table.clear();
}

bool GoParserBase::closingBracket()
{
    antlr4::BufferedTokenStream* stream = static_cast<antlr4::BufferedTokenStream*>(_input);
    auto la = stream->LT(1);
    return la->getType() == GoParser::R_CURLY || la->getType() == GoParser::R_PAREN || la->getType() == antlr4::Token::EOF;
}

bool GoParserBase::isNotReceive()
{
    antlr4::BufferedTokenStream* stream = static_cast<antlr4::BufferedTokenStream*>(_input);
    auto la = stream->LT(2);
    return la->getType() != GoParser::RECEIVE;
}

std::vector<std::string> split(const std::string& str, char delimiter)
{
    std::vector<std::string> tokens;
    std::stringstream ss(str);
    std::string token;

    while (std::getline(ss, token, delimiter))
    {
        tokens.push_back(token);
    }

    return tokens;
}

void GoParserBase::addImportSpec()
{
    antlr4::ParserRuleContext* ctx = this->_ctx;
    auto importSpec = dynamic_cast<GoParser::ImportSpecContext*>(ctx);
    if (importSpec == nullptr) return;
    auto packageName = importSpec->packageName();
    if (packageName != nullptr)
    {
        auto name = packageName->getText();
        if (debug) std::cout << "Entering " << name << std::endl;
        table.insert(name);
        return;
    }
    auto importPath = importSpec->importPath();
    if (importPath == nullptr) return;
    auto name = importPath->getText();
    if (debug) std::cout << "import path " << name << std::endl;
    name.erase(std::remove(name.begin(), name.end(), '\"'), name.end());
    if (name.empty()) return;
    std::replace(name.begin(), name.end(), '\\', '/');
    auto pathArr = split(name, '/');
    if (pathArr.empty()) return;
    auto lastComponent = pathArr[pathArr.size()-1];
    if (lastComponent.empty()) return;
    // Handle special cases like "." and ".."
    if (lastComponent == "." || lastComponent == "..") return;
    auto fileArr = split(lastComponent, '.');
    // Guard against empty array (can happen if lastComponent is all dots)
    if (fileArr.empty()) {
        table.insert(lastComponent);
        if (debug) std::cout << "Entering " << lastComponent << std::endl;
        return;
    }
    auto fileName = fileArr[fileArr.size()-1];
    if (fileName.empty()) {
        // Fall back to lastComponent if split resulted in empty string
        fileName = lastComponent;
    }
    if (debug) std::cout << "Entering " << fileName << std::endl;
    table.insert(fileName);
}

bool GoParserBase::isOperand()
{
    antlr4::BufferedTokenStream* stream = static_cast<antlr4::BufferedTokenStream*>(_input);
    auto la = stream->LT(1);
    if (la->getText() == "err") return true;
    bool result = true;
    if (la->getType() != GoParser::IDENTIFIER) {
        if (debug) std::cout << "isOperand Returning " << result << " for " << la << std::endl;
        return result;
    }
    result = table.find(la->getText()) != table.end();
    auto la2 = stream->LT(2);
    // If it's not followed by a '.', then it really should be
    // considered as operand.
    if (la2->getType() != GoParser::DOT) {
        result = true;
        if (debug) std::cout << "isOperand Returning " << result << " for " << la << std::endl;
        return result;
    }
    // If it's followed by '.', and then followed by '(', then
    // it is a typeAssertion, and so la must be an operand.
    auto la3 = stream->LT(3);
    if (la3->getType() == GoParser::L_PAREN)
    {
        result = true;
        if (debug) std::cout << "isOperand Returning " << result << " for " << la << std::endl;
        return result;
    }
    if (debug) std::cout << "isOperand Returning " << result << " for " << la << std::endl;
    return result;
}

bool GoParserBase::isConversion()
{
    antlr4::BufferedTokenStream* stream = static_cast<antlr4::BufferedTokenStream*>(_input);
    auto la = stream->LT(1);
    auto result = la->getType() != GoParser::IDENTIFIER;
    if (debug) std::cout << "isConversion Returning " << result << " for " << la << std::endl;
    return result;
}

bool GoParserBase::isMethodExpr()
{
    antlr4::BufferedTokenStream* stream = static_cast<antlr4::BufferedTokenStream*>(_input);
    auto la = stream->LT(1);
    bool result = true;
    // See if it looks like a method expr.
    if (la->getType() == GoParser::STAR) {
        if (debug) std::cout << "isMethodExpr Returning " << result << " for " << la << std::endl;
        return result;
    }
    if (la->getType() != GoParser::IDENTIFIER) {
        result = false;
        if (debug) std::cout << "isMethodExpr Returning " << result << " for " << la << std::endl;
        return result;
    }
    result = ! (table.find(la->getText()) != table.end());
    if (debug) std::cout << "isMethodExpr Returning " << result << " for " << la << std::endl;
    return result;
}

// Built-in functions that take a type as first argument
const std::set<std::string> GoParserBase::BUILTIN_TYPE_FUNCTIONS = {"make", "new"};

// Check if we're in a call to a built-in function that takes a type as first argument.
// Called after L_PAREN has been matched in the arguments rule.
bool GoParserBase::isTypeArgument()
{
    antlr4::BufferedTokenStream* stream = static_cast<antlr4::BufferedTokenStream*>(_input);
    // After matching L_PAREN, LT(-1) is '(' and LT(-2) is the token before it
    auto funcToken = stream->LT(-2);
    if (funcToken == nullptr || funcToken->getType() != GoParser::IDENTIFIER) {
        if (debug) std::cout << "isTypeArgument Returning false - no identifier before (" << std::endl;
        return false;
    }
    bool result = BUILTIN_TYPE_FUNCTIONS.find(funcToken->getText()) != BUILTIN_TYPE_FUNCTIONS.end();
    if (debug) std::cout << "isTypeArgument Returning " << result << " for " << funcToken->getText() << std::endl;
    return result;
}

// Check if we're NOT in a call to a built-in function that takes a type.
// This is the inverse of isTypeArgument for the expressionList alternative.
bool GoParserBase::isExpressionArgument()
{
    bool result = !isTypeArgument();
    if (debug) std::cout << "isExpressionArgument Returning " << result << std::endl;
    return result;
}

